pywhy_stats/kci.py

```diff
@@ -0,0 +1,374 @@
+    kernel_Z : Callable expecting a numpy array as input and returns another numpy array
+        The kernel function for Z. By default, the RBF kernel is used for continuous and the delta kernel for
+        categorical data.
+    approx : bool
+        Whether to use the Gamma distribution approximation for the pvalue, by default True.
+    null_sample_size : int
+        The number of samples to generate for the bootstrap distribution to approximate the pvalue, by default 1000.
+    threshold : float
+        The threshold set on the value of eigenvalues, by default 1e-5. Used to regularize the method.
+    normalize_data : bool
+        Whether the data should be standardized to unit variance, by default True.
+
+    Notes
+    -----
+    Any callable can be given to create the kernel matrix. For instance, to use a particular kernel from sklearn:
+        kernel_X = func:`sklearn.metrics.pairwise.pairwise_kernels.polynomial`
+
+    References
+    ----------
+    .. footbibliography::
+    """
+    test_statistic, pvalue = _kernel_test(X, Y, Z, kernel_X, kernel_Y, kernel_Z, approx, null_sample_size, threshold,
+                                          normalize_data)
+    return PValueResult(pvalue=pvalue, statistic=test_statistic)
+
+
+def _kernel_test(
+        X: ArrayLike,
+        Y: ArrayLike,
+        Z: Optional[ArrayLike],
+        kernel_X: Optional[Callable[[np.ndarray], np.ndarray]],
+        kernel_Y: Optional[Callable[[np.ndarray], np.ndarray]],
+        kernel_Z: Optional[Callable[[np.ndarray], np.ndarray]],
+        approx: bool,
+        null_sample_size: int,
+        threshold: float,
+        normalize_data: bool
+) -> Tuple[float, float]:
+    if X.ndim == 1:
+        X = X.reshape(-1, 1)
+    if Y.ndim == 1:
+        Y = Y.reshape(-1, 1)
+    if Z is not None and Z.ndim == 1:
+        Z = Z.reshape(-1, 1)
+
+    if normalize_data:
+        # first normalize the data to have zero mean and unit variance
+        # along the columns of the data
+        X = _normalize_data(X)
+        Y = _normalize_data(Y)
+        if Z is not None:
+            Z = _normalize_data(Z)
+
+    if kernel_X is None:
+        kernel_X = _get_default_kernel(X)
+    if kernel_Y is None:
+        kernel_Y = _get_default_kernel(Y)
+    if Z is not None and kernel_Z is None:
+        kernel_Z = _get_default_kernel(Z)
+
+    Kx = kernel_X(X)
+    Ky = kernel_Y(Y)
+    if Z is not None:
+        Kz = kernel_Z(Z)
+        Kx = Kx * Kz  # Equivalent to concatenating them beforehand. However, here we can then have individual kernels.
+        Kz = _fast_centering(Kz)
+
+    Kx = _fast_centering(Kx)
+    Ky = _fast_centering(Ky)
+
+    if Z is None:
+        return _ind(Kx, Ky, approx, null_sample_size, threshold)
+    else:
+        return _cond(Kx, Ky, Kz, approx, null_sample_size, threshold)
+
+
+def _ind(Kx, Ky, approx, null_sample_size, threshold):
+    # test statistic is just the normal bivariate independence
+    # test statistic
```

Comment:
```suggestion
    # test statistic is the normal bivariate independence test
```

? What is this meant to say?